### PR TITLE
add disable archiving segment for plugins

### LINKED ISSUE #18134: Add new configuration to disable archiving the segment reports for specific plugins
We noticed some plugins may cause a lot of CPU load. A solution is to disable a plugin that causes a lot of load. However, sometimes it be great to still have at least report data for "All visits" but not segments. This way say 95% of the load is reduced while the biggest value might be still there.

It be great if we had a new config ini flag to disable archiving the archiving for some plugins when the archiving is for any segment. 

When this is disabled for a plugin then:

* Don't archive any data for this plugin if segment is not empty
* In the report itself, when a segment is viewed and there is no data, then show a footer message in the report that viewing segments for this report is disabled.
* Reports for "all visits" (segment is empty) should still be archived regularly

Basically, a config setting like 

```
[General]
disable_archiving_segment_for_plugin="Plugin1,Plugin2,Plugin3"
```

Where we can provide the name of plugins in a comma separated list. Could also use an array if that makes it easier like
 
```
[General]
disable_archiving_segment_for_plugin[]="Plugin1"
disable_archiving_segment_for_plugin[]="Plugin1"
disable_archiving_segment_for_plugin[]="Plugin1"
```

The config name is just a random thought without too much thinking. Feel free to change the name.

In tracking we can configure some settings on a per site basis see https://matomo.org/faq/how-to/how-do-i-configure-a-config-setting-in-the-config-ini-php-differently-for-a-specific-site/

If we could do this for this setting as well that would be even better. Eg if we could set below setting to only apply this for site 42 that would be even better.

```
[General_42]
disable_archiving_segment_for_plugin[]="Plugin1"
```

### PR DESCRIPTION
**Pull request by peterhashair**
### Description:
Fixes: #18134
add disable archiving segment configuration for plugins

- Under `[General]`-> `disable_archiving_segment_for_plugins` accept string, sting with comma, array. 
- `[General_{siteId}]` can be enable by Site ID. 

I guess this will be 4.7

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: config/global.ini.php
Comment (justinvelluppillai) [inline config/global.ini.php]: Maybe better "This field accepts more than one plugin separated by comma eg:..."

File: config/global.ini.php
Comment (justinvelluppillai) [inline config/global.ini.php]: Should be "accepts"

File: config/global.ini.php
Comment (justinvelluppillai) [inline config/global.ini.php]: "if you'd like to disable on a specific site"

File: config/global.ini.php
Comment (bx80) [inline config/global.ini.php]:
```suggestion
; if you'd like to disable on a specific site, please append the siteId to the [General_siteId]. eg siteId=2 [General_2]
```

**Review by bx80 — APPROVED** [review]

**Comment by peterhashair** [comment]
The last test runs fully passed [https://app.travis-ci.com/github/matomo-org/matomo/builds/241828886](https://app.travis-ci.com/github/matomo-org/matomo/builds/241828886). the Last update just one wording in global.config.php, I think no need to run the test again.

**Comment by justinvelluppillai** [comment]
@peterhashair changing the config changes one of the screenshots

**Comment by peterhashair** [comment]
@justinvelluppillai right, let me run it again.

**Comment by sgiehl** [comment]
> [General_{siteId}] can be enable by Site ID.

If I see that right, the site specific config would then only work for that specific configuration value, right? Imho we should do it the same way like in tracker config. So it would be available for all values in General section.

See https://github.com/matomo-org/matomo/blob/e0417845f8fd7031e00e97c3576415c998abb2ce/core/Tracker/TrackerConfig.php#L28-L47

I'm not sure if it's easy to adjust the config class to support that. Maybe we need to introduce a new class `GeneralConfig` (like `TrackerConfig`) and replace all usages of the general config over time then...

**Comment by peterhashair** [comment]
@sgiehl  right, that makes sense let me do the adjustment.

File: core/ArchiveProcessor/Rules.php
Comment (sgiehl) [inline core/ArchiveProcessor/Rules.php]: Shouldn't this part now be obsolete? The `GeneralConfig` below should now return either the site specific or the global config

File: config/global.ini.php
Comment (sgiehl) [inline config/global.ini.php]: If we aim to support this feature for all settings in General config, we maybe don't need to mention it in each setting. 🤔

**Review by sgiehl — CHANGES_REQUESTED** [review]

**Comment by peterhashair** [comment]
@sgiehl yes, don't need load `General`. sorry, my brain was dead there 😅

File: config/global.ini.php
Comment (peterhashair) [reply config/global.ini.php]: @sgiehl do we want to remove this part?

Comment (sgiehl) [reply config/global.ini.php]: Might be better to mention that somewhere else, like in the FAQ

Comment (peterhashair) [reply config/global.ini.php]: @sgiehl right that makes sense, let me convert to FAQ tomorrow. 👍

**Comment by peterhashair** [comment]
FAQ for review https://matomo.org/?post_type=faq&p=47059&preview=true

File: config/global.ini.php
Comment (sgiehl) [inline config/global.ini.php]:
```suggestion
; Allows you to disable archiving segments for selected plugins. For more details please see https://matomo.org/faq/how-to-disable-archiving-the-segment-reports-for-specific-plugins
```

File: core/ArchiveProcessor/Rules.php
Comment (sgiehl) [inline core/ArchiveProcessor/Rules.php]: This can be simplified as:

```suggestion
        return in_array($pluginName, $pluginArchivingSetting);
```

File: tests/PHPUnit/Integration/ArchiveProcessor/LoaderTest.php
Comment (sgiehl) [inline tests/PHPUnit/Integration/ArchiveProcessor/LoaderTest.php:1491]: Might be good to also add a test if overwriting works correctly. e.g. a config like

```
[General]
disable_archiving_segment_for_plugins = myplugin

[General_1]
disable_archiving_segment_for_plugins = testplugin
```

Guess for idSite=1 only `testplugin` should then be skipped and not `myplugin`

File: core/ArchiveProcessor/Rules.php
Comment (sgiehl) [inline core/ArchiveProcessor/Rules.php]: @tsteur should we compare the plugin name case insensitive here? Or should that only work if the plugin name is exactly the same?

**Review by sgiehl — CHANGES_REQUESTED** [review]
Left a couple of minor comments, afterwards I guess this should be ready to merge

File: core/ArchiveProcessor/Rules.php
Comment (tsteur) [reply core/ArchiveProcessor/Rules.php]: No big preference. Maybe case insensitive be better if done easily.

**Comment by peterhashair** [comment]
@sgiehl update a little bit, 
-  add site Id configure test overwrite general test. 
- case insensitive, 
- special case ignored.

**Comment by peterhashair** [comment]
@sgiehl add test if overwriting to here  [https://github.com/matomo-org/matomo/blob/d6b4e69967233a8181e23bfa2295980fbfb26614/tests/PHPUnit/System/ArchiveInvalidationTest.php#L198-L201](https://github.com/matomo-org/matomo/blob/d6b4e69967233a8181e23bfa2295980fbfb26614/tests/PHPUnit/System/ArchiveInvalidationTest.php#L198-L201)

File: core/ArchiveProcessor/Rules.php
Comment (sgiehl) [inline core/ArchiveProcessor/Rules.php]: Won't that also return partial matches. e.g `testplugin` would also match if `testplugin2` would be disabled, right?

File: core/ArchiveProcessor/Rules.php
Comment (sgiehl) [inline core/ArchiveProcessor/Rules.php]: To check for valid plugin name there is also `Plugin\Manager::isValidPluginName`

File: tests/PHPUnit/System/ArchiveInvalidationTest.php
Comment (sgiehl) [inline tests/PHPUnit/System/ArchiveInvalidationTest.php:202]: Maybe a test to ensure no partial matches work would be good:

```suggestion

        Config::setSetting('General_1', 'disable_archiving_segment_for_plugins', 'testPlugin2');
        $this->assertFalse(Rules::isSegmentPluginArchivingDisabled('testPlugin', 1));
```

File: core/ArchiveProcessor/Rules.php
Comment (sgiehl) [inline core/ArchiveProcessor/Rules.php]: To fix the previous mentioned problems I would suggest to use something like this:

```suggestion
    public static function isSegmentPluginArchivingDisabled($pluginName, $siteId = null)
    {
        $pluginArchivingSetting = GeneralConfig::getConfigValue('disable_archiving_segment_for_plugins', $siteId);

        if (empty($pluginArchivingSetting)) {
            return false;
        }

        if (is_string($pluginArchivingSetting)) {
            $pluginArchivingSetting = explode(",", $pluginArchivingSetting);
            $pluginArchivingSetting = array_filter($pluginArchivingSetting, function($plugin){
                return Manager::getInstance()->isValidPluginName($plugin);
            });
        }

        $pluginArchivingSetting = array_map('strtolower', $pluginArchivingSetting);

        return in_array(strtolower($pluginName), $pluginArchivingSetting);
    }
```

**Review by sgiehl — CHANGES_REQUESTED** [review]

**Comment by peterhashair** [comment]
@sgiehl that make more sense. 👍

**Review by sgiehl — DISMISSED** [review]
Should be good to merge now if tests are passing.

**Comment by sgiehl** [comment]
@peterhashair Just had another look at the original issue and it seems there is one requirement, that isn't fulfilled currently:

> In the report itself, when a segment is viewed and there is no data, then show a footer message in the report that viewing segments for this report is disabled.

If you need some hints to find the correct places to implement this. Maybe start having a look here:
https://github.com/matomo-org/matomo/blob/5d9cf649aeefa7a1390574c2b147ce972b1b5c39/core/Plugin/Visualization.php#L228-L235
https://github.com/matomo-org/matomo/blob/1023fb7d691ebf47c27aa8a1cab4702c55c2630d/plugins/CoreHome/templates/_dataTable.twig#L58-L68

**Comment by peterhashair** [comment]
@sgiehl I did a mock-up for the requirement. Not sure that's the right approach for this, do I need to implement a UI test for that one?

> In the report itself, when a segment is viewed and there is no data, then show a footer message in the report that viewing segments for this report are disabled.

File: core/Plugin/Visualization.php
Comment (sgiehl) [inline core/Plugin/Visualization.php]:
```suggestion
```

File: plugins/CoreHome/lang/en.json
Comment (sgiehl) [inline plugins/CoreHome/lang/en.json]: We should try to write messages that are useful for all kind of users. This one might be a bit too technical and the FAQ link might only be relevant for the admin. Guess something simple like `Segmentation is currently disabled for this report` would do it. Not sure if we maybe should mention to contact the admin for details 🤔 Maybe @tsteur has an even better suggestion for that.

And a general thing: We should not include an URL in a translation, as that might be error prone for translations. In that case you could have only used one placeholder to include everything, or link a text like `this FAQ` instead.

File: core/Plugin/Visualization.php
Comment (sgiehl) [inline core/Plugin/Visualization.php]: You also need to check if a segment was actually applied to this report, Otherwise it will also be shown for unsegmented reports, that have no data for other reasons.

File: plugins/CoreHome/templates/_dataTable.twig
Comment (sgiehl) [inline plugins/CoreHome/templates/_dataTable.twig]: Might be better to check if the variable is defined, otherwise it might throw an exception if that is not the case.
```suggestion
                {% elseif showPluginArchiveDisabled is defined and showPluginArchiveDisabled %}
```

File: plugins/CoreHome/templates/_dataTable.twig
Comment (sgiehl) [inline plugins/CoreHome/templates/_dataTable.twig]: In case you output a translation in a twig template that contains html, you need to apply the `raw` filter, otherwise the HTML is outputted encoded.

**Review by sgiehl — COMMENTED** [review]
@peterhashair your approach has the correct direction ;-)

File: core/Config/GeneralConfig.php
Comment (sgiehl) [inline core/Config/GeneralConfig.php]: Could improve this, to avoid random array access errors when we start using this class more.
```suggestion
        return $config[$name] ?? null;
```

File: config/global.ini.php
Comment (sgiehl) [inline config/global.ini.php:389]: Wouldn't it be good to directly mention that it's an comma separated list?

File: plugins/CoreHome/lang/en.json
Comment (sgiehl) [inline plugins/CoreHome/lang/en.json]:
```suggestion
        "DataForThisReportHasBeenDisabled": "Segmentation is currently disabled for this report. Please check %1$sthis FAQ%2$s for more details.",
```

File: plugins/CoreHome/templates/_dataTable.twig
Comment (sgiehl) [inline plugins/CoreHome/templates/_dataTable.twig]: The escaping of the link would actually not be needed, as it won't break the html. This is only required for variables were we don't now what content it might add. But as it doesn't hurt, there is no need to change that.

**Review by sgiehl — CHANGES_REQUESTED** [review]
@peterhashair Left a couple of comments for smaller improvements. Regarding a UI test: I guess it would be good to have one, as it would also be kind of an integration test, as it ensures that tracked data isn't archived and displayed. Besides that guess it should be ready to merge after the adjustments

File: tests/PHPUnit/Fixtures/DisablePluginArchive.php
Comment (sgiehl) [inline tests/PHPUnit/Fixtures/DisablePluginArchive.php]: It would be good to also track some visits in this fixture, maybe one or two with different referrers. That would ensure that no archiving is done. Otherwise the reports would show the data...

**Review by sgiehl — CHANGES_REQUESTED** [review]
@peterhashair it seems you have updated a couple of screenshots that are unrelated to this issue. The updater screenshots are failing for a specific reason, and they should work again, once that issue is fixed. So the changes to them should be reverted. Also you added `tests/UI/expected-screenshots/DisablePluginArchive.png`, which is not used.

**Comment by peterhashair** [comment]
@sgiehl right, update below. I am not too sure the refer tracking is correct, copied from other fixtures. 
- remove the unrelated UI changes.
- set up a refer tracking.
- update screenshot name

File: plugins/CoreHome/lang/en.json
Comment (sgiehl) [inline plugins/CoreHome/lang/en.json:14]: @tsteur Not sure if this will be a feature we are going to use on cloud maybe. In that case I was wondering if linking an FAQ in the message would actually be helpful, as the user can't change that 🤔

**Review by sgiehl — APPROVED** [review]
The remaining failing tests seem unrelated, so should be good to merge now. Left a comment regarding cloud, so might be good to wait for an answer.

**Comment by tsteur** [comment]
@sgiehl linking to an FAQ be great, the FAQ could also mention specifically the cloud that if they are on the cloud, to reach out to our support.

**Comment by sgiehl** [comment]
@peterhashair I have merged in the latest changes from `4.x-dev`. If tests are passing, feel free to merge

**Pull request merged**
